engine/tls.go

```diff
@@ -0,0 +1,234 @@
+type TLSSettings struct {
+	// PreferServerCipherSuites controls whether the server selects the CipherSuites
+	PreferServerCipherSuites bool
+
+	// SkipVerify skips certificate check, very insecure
+	InsecureSkipVerify bool
+
+	// MinVersion is minimal TLS version "VersionTLS10" is default
+	MinVersion string
+
+	// MaxVersion is max supported TLS version, "VersionTLS12" is default
+	MaxVersion string
+
+	// SessionTicketsDisabled disables session ticket resumption support
+	SessionTicketsDisabled bool
+
+	// SessionCache specifies TLS session cache, default it 'LRU' with capacity 1024
+	SessionCache TLSSessionCache
+
+	// Preferred CipherSuites, default is
+	CipherSuites []string
+}
+
+// TLSSessionCache sets up parameters for TLS session cache
+type TLSSessionCache struct {
+	// Session cache implementation, default is 'LRU' with capacity 1024
+	Type     string
+	Settings *LRUSessionCacheSettings
+}
+
+type LRUSessionCacheSettings struct {
+	// LRU Capacity default is 1024
+	Capacity int
+}
+
+// NewTLSConfig validates the TLSSettings and returns the tls.Config with the converted parameters
+func NewTLSConfig(s *TLSSettings) (*tls.Config, error) {
+	// Parse min and max TLS versions
+	var min, max uint16
+	var err error
+
+	if s.MinVersion == "" {
+		min = tls.VersionTLS10
+	} else if min, err = ParseTLSVersion(s.MinVersion); err != nil {
+		return nil, err
+	}
+
+	if s.MaxVersion == "" {
+		max = tls.VersionTLS12
+	} else if max, err = ParseTLSVersion(s.MaxVersion); err != nil {
+		return nil, err
+	}
+
+	var css []uint16
+	if len(s.CipherSuites) == 0 {
+		css = []uint16{
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+
+			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
+
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
+
+			tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_RSA_WITH_AES_128_CBC_SHA,
+		}
+	} else {
+		css = make([]uint16, len(s.CipherSuites))
+		for i, suite := range s.CipherSuites {
+			cs, err := ParseCipherSuite(suite)
+			if err != nil {
+				return nil, err
+			}
+			css[i] = cs
+		}
+	}
+
+	var cache tls.ClientSessionCache
+	if !s.SessionTicketsDisabled {
+		cache, err = NewTLSSessionCache(&s.SessionCache)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &tls.Config{
+		MinVersion: min,
+		MaxVersion: max,
+
+		SessionTicketsDisabled: s.SessionTicketsDisabled,
+		ClientSessionCache:     cache,
+
+		PreferServerCipherSuites: s.PreferServerCipherSuites,
+		CipherSuites:             css,
+
+		InsecureSkipVerify: s.InsecureSkipVerify,
+	}, nil
+}
+
+// NewTLSSessionCache validates parameters and creates a new TLS session cache
+func NewTLSSessionCache(s *TLSSessionCache) (tls.ClientSessionCache, error) {
+	cacheType := s.Type
+	if cacheType == "" {
+		cacheType = LRUCacheType
+	}
+	if cacheType != "LRU" {
```

Comment:
Can you refer to LRU the same way here and two lines above? Either `LRUCacheType` or `"LRU"`. I somewhat prefer `"LRU"`, but either way is fine.